### PR TITLE
Kalman BMTF Emu v2.7

### PR DESCRIPTION
**Pull request by panoskatsoulis**
This is the latest BMTF emulator for the Kalman algorithm

### CODE REVIEW

**Comment by thomreis** [comment]
please test

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
since the master points to the 10_4 branch now please also make a backport PR to 10_3_X after fixing the compilation warnings.

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
please fix the compilation warnings in order to go ahead with this PR's integration.

**Comment by panoskatsoulis** [comment]
hi @thomreis 
the procedure fails because of 2 Clang warnings (namely **-Wself-assign-overloaded** & **-Wheader-guard**).
The last commit I've pushed fixes them in my local tests.

In details, -Wheader-guard is just a typo in the first line of the file
'L1Trigger/L1TMuonBarrel/interface/L1TMuonBarrelKalmanLUTs.h',
but -Wself-assign-overloaded raised during compilation because of the line 226 in the file
'L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelKalmanRegionModule.cc'.
**line 226:   s3_3=s3_3;**
The assignment of an object to itself now raises this warning.
In the release notes of the Clang ([link](http://ftp.netbsd.org/pub/NetBSD/NetBSD-current/src/external/bsd/llvm/dist/clang/docs/ReleaseNotes.rst)) is proposed to cast the rhs as reference of the object, in order to suppress the warning.
**line 226:   s3_3=static_cast<L1MuKBMTrack&>(s3_3);**

Can we please test again and see if this fixes the issue?
Also, @bachtis confirm please that this fix will not affect the algorithm.

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
what is the point of assigning an object to itself? I can see the point of having this in FW but not in an emulator.

**Comment by panoskatsoulis** [comment]
I would say possibly for following the same steps exactly as in FW (the difference of the emulation concept wrt the simulation concept).
I think there is a reason why @bachtis is doing this.

**Comment by thomreis** [comment]
OK, so the @bachtis can confirm if this is still compatible with the FW.
In the meantime we can already test this.
Please also update the 10_2 backport and create a 10_3_X one when the fix is confirmed.

**Comment by thomreis** [comment]
please test

**Comment by panoskatsoulis** [comment]
sure, I'll update the other PR while we wait for answer and the test results

**Comment by bachtis** [comment]
Guys wait. Maybe CLANG found a bug here . Looking to it

**Comment by bachtis** [comment]
OK looked at it. @panoskatsoulis  please replace the line 226 to      
s3_3=s2_3; 
and update the PR .
In short: This code is a mimicing of the hardware sorter  in firmware so it is more complicated than a std::sort for data/MC agreement to work . The line of self assesment is a typo .

**Comment by thomreis** [comment]
please test

**Comment by panoskatsoulis** [comment]
good, the typo has been corrected.
Although it wont correct the RelVal issues I think.
@thomreis please wait for the test, I need to check the logs

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
are the failing histograms expected with this emulator change?

**Comment by thomreis** [comment]
Please provide a PR to 10_3_X as well.

**Comment by panoskatsoulis** [comment]
Hi all,
The failing plots belong all in the "Secondary" L1TEMU/L1TdeStage2BMTF folder.
For the run 297557 (is used by the workflow 136.788) in this folder are the Kalman Algorithm Validation plots.
However, the firmware which is used for this run doesn't include any Kalman Algorithm at all.
The Unpacker in this case creates a null collection that is being compared to the Kalman output giving a validation that is completely without sense.

Maybe I need to understand what are the data points that lay on these plots in such cases (because they might be leftovers in the memory) but this is completely irrelevant to the validity of this PR.

The failing plots are not related with the PR'ed Emulator.

**Comment by thomreis** [comment]
+1

**Comment by panoskatsoulis** [comment]
hold

**Comment by panoskatsoulis** [comment]
unhold

**Comment by thomreis** [comment]
please test

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
are the failing comparisons expected?

**Comment by panoskatsoulis** [comment]
Hi @thomreis 
which are failing comparisons? Can't find them

**Comment by thomreis** [comment]
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_4_X_2018-10-18-2300+24855/28994/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/L1TEMU_L1TdeStage2BMTF_L1TdeStage2BMTF-Secondary.html for example but there could be others.

**Comment by panoskatsoulis** [comment]
@thomreis 
I've checked the failures. I would say they are not problem.
The test is checking consistency of the master release with and without the PR for the emulator plots.
The emulator plots without the PR are wrong so this is why we pushed this. The latest (PR version) emulator and the deprecated (master version) emulator is expected that they won't agree in their mismatches.

There is no way to have consistency of the mismatches with the new and the old Emulator I think.
The failures, all, seem to come from the "Secondary" folder where the Kalman comparison is taking place. Did you find any different?

**Comment by thomreis** [comment]
+1

@panoskatsoulis thanks for checking.

File: L1Trigger/L1TMuonBarrel/test/kalmanTools/makeEtaLuts.py [hunk 1]
```diff
@@ -1,4 +1,3 @@
-from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (bachtis) [reply L1Trigger/L1TMuonBarrel/test/kalmanTools/makeEtaLuts.py]: I didnt add this piece in the code. probably someone did it to make it compile in different python version. Just delete the line

**Comment by fabiocos** [comment]
@panoskatsoulis @bachtis all the python fragments must be python3 compatible, and @davidlange6 made a CMSSW-wide migration for this a few months ago. So deleting that line is not an option, you should just move the syntax of the print statement to the new one. It should be a fairly straightforward adjustment.

**Comment by fabiocos** [comment]
hold

pending the required fixes

**Comment by fabiocos** [comment]
@panoskatsoulis any news? the requested update looks trivial to be implemented

**Comment by panoskatsoulis** [comment]
Hi @thomreis , @fabiocos , sorry for the delay... I will update it in a few days.

**Comment by panoskatsoulis** [comment]
Hi, again @thomreis, @fabiocos 
i've redone the python3 migration, and also changed all print statements to the python3 print() function

**Comment by thomreis** [comment]
please test

**Comment by panoskatsoulis** [comment]
@fabiocos,
will it be needed to make the same change to PR #24934 as well?
That one is related to 103X.

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
yes.

**Comment by fabiocos** [comment]
@thomreis if the backport to 10_3_X is really needed (is it now at the end of HI data taking ?) the code should be updated accordingly.

**Comment by thomreis** [comment]
Hi @panoskatsoulis can you check if all the failing comparisons are expected:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_4_X_2018-11-25-2300+24855/29530/validateJR.html
Please take a look also at the static checks https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24855/31849/llvm-analysis/

**Comment by panoskatsoulis** [comment]
Hi @thomreis 
after talking with @bachtis, It seems that some lines from the src were of no need.
I've updated the PR once again.

**Comment by thomreis** [comment]
please test

If the lines are not needed it would be better to remove them completely.

**Comment by panoskatsoulis** [comment]
It is possible the commented out lines to be useful under debugging circumstances. I wouldn't remove them without @bachtis 's input on the issue.

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
are the differences in the test WFs expected?

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by panoskatsoulis** [comment]
The following latest changes are the problem, I will update it again to be compatible with the 10_4_X interface.

<<<<<<< BMTF_KMTF-Emu-v2.7_forMaster
#ifndef __L1TMuonBarrelKalmanLUTs_h
#define __L1TMuonBarrelKalmanLUTs_h
=======
#ifndef L1Trigger_L1TMuonBarrel_L1TMuonBarrelKalmanLUTs_h
#define L1Trigger_L1TMuonBarrel_L1TMuonBarrelKalmanLUTs_h
>>>>>>> master

**Comment by panoskatsoulis** [comment]
Hi @thomreis
the failures seem to be because of the old runs which have been used in the workflows, and the non-well-defined behavior of the DQM AlgoSelector plugin under these runs.
There is a default behavior of the "algoselector" to prevent segfaults when is processing runs in which there was no Kalman firmware running at the time.
However, this plugin has been working correctly with the latest runs at P5 for more than 2 months and also is not part of this PR.

**Comment by thomreis** [comment]
please test

**Comment by thomreis** [comment]
please test

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
there is a violation of CMS code rules https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24855/32121/codeRules/
Please take a look.

**Comment by thomreis** [comment]
In the meantime @fabiocos could you unhold this already please? The python fixes have been implemented already a while ago.

**Comment by thomreis** [comment]
Hi @panoskatsoulis 
false alarm. The code rule violation must have been in a commit that was added to the tests but is not in this PR.
Please check the mismatches and confirm again that they are all expected.

**Comment by fabiocos** [comment]
unhold

**Comment by thomreis** [comment]
+1

The failing comparisons are the same than in previous tests and due to the use of old runs in the WFs where the Kalman FW was not active in data taking.

**Comment by panoskatsoulis** [comment]
Hi @thomreis, I've checked again. In deed the failures exist due to the old runs used by the workflow, and also we need to keep in mind that this PR changes the emulator. With that in mind, differences are being expected even for runs in which the emulator run.

**Comment by thomreis** [comment]
Hi @panoskatsoulis thanks for the additional confirmation.
@fabiocos this is finally ready to be merged.

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
@thomreis @panoskatsoulis the PR looks ready for integration, just one clarification: I see several pieces of code that are not just debugging printouts commented: what is the purpose of that? In principle this might turn into a time bomb if not properly documented.

**Comment by panoskatsoulis** [comment]
Hello, @fabiocos this question is for @bachtis .

**Comment by bachtis** [comment]
Hi Fabio, We will clean it up some time in the future - but I suggest we merge this now. We will have additional developments and PRs in the near future so we can clean it up .

**Comment by fabiocos** [comment]
+1

**Pull request merged**